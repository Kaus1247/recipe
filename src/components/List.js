import PropTypes from 'prop-types';
import styled from 'styled-components';
import { space, layout } from 'styled-system';
import propTypes from '@styled-system/prop-types';

export const ListItem = styled.li`
  padding: ${(props) => {return props.padding || 0 }};
  border-top: ${(props) => {return props.borderTop || 0 }};
`;

ListItem.defaultProps = {
  margin: 0,
  padding: 0,
  width: '100%',
  border: 'none',
};

ListItem.displayName = 'ListItem';

const List = styled.ul`
  list-style-type: ${(props) => props.listStyleType};
  ${space}
  ${layout}
`;

List.propTypes = {
  ...propTypes.layout,
  ...propTypes.space,
  listStyleType: PropTypes.string,
};

List.defaultProps = {
  listStyleType: 'none',
  margin: 0,
  padding: 0,
  width: '100%',
};

List.displayName = 'List';

export default List;
