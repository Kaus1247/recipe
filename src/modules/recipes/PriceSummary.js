import React from 'react';
import Box from '../../components/Box';
import { Row, Col } from '../../components/Grid';
import Flex from '../../components/Flex';
import List, { ListItem } from '../../components/List';
import Text from '../../components/Text';
import { parseRawPrice } from './price';

// Create PriceSummary user interface
const PriceSummary = ({ summary = (new Map()), totalPrice="$0.00", shippingPrice=0 }) => {
  return (
    <Box width={['290px', '450px']}>
      {
        summary.size > 0
          ? <List padding="16px">
              {
                Array.from(summary.values()).map((recipe, index)=>{
                  return (
                    <ListItem key={index+recipe.id} padding="0px 0px 8px 0px">
                      <Row>
                        <Col sm={9}>
                          <Flex alignItems="center" justifyContent="flex-start">
                            <Text fontWeight="400"> { recipe.name } x { recipe.selected }</Text>
                          </Flex>
                        </Col>
                        <Col sm={3}>
                          <Flex alignItems="center" justifyContent="flex-end">
                            <Text fontWeight="400">{ parseRawPrice(recipe.totalRecipePrice) }</Text>
                          </Flex>
                        </Col>
                      </Row>
                    </ListItem>
                  )
                })
              }
              <ListItem key={"shippingPrice"} padding="0px 0px 8px 0px">
                <Row>
                  <Col sm={9}>
                    <Flex alignItems="center" justifyContent="flex-start">
                      <Text fontWeight="400"> Shipping </Text>
                    </Flex>
                  </Col>
                  <Col sm={3}>
                    <Flex alignItems="center" justifyContent="flex-end">
                      <Text fontWeight="400">{ parseRawPrice(shippingPrice) }</Text>
                    </Flex>
                  </Col>
                </Row>
              </ListItem>
              <ListItem key={"totalBoxPrice"} padding="8px 0px 0px 0px" borderTop="1px solid #E4E4E4">
                <Row>
                  <Col sm={9}>
                    <Flex alignItems="center" justifyContent="flex-start">
                      <Text fontWeight="600"> Total </Text>
                    </Flex>
                  </Col>
                  <Col sm={3}>
                    <Flex alignItems="center" justifyContent="flex-end">
                      <Text fontWeight="600">{ totalPrice }</Text>
                    </Flex>
                  </Col>
                </Row>
              </ListItem>
            </List>
          : null
      }

    </Box>
  )
};

export default PriceSummary;
