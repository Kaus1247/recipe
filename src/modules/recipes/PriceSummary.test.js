import React from "react";
import PriceSummary from "./PriceSummary";
import { shallow } from "enzyme";
import List, { ListItem } from '../../components/List';
import Box from "../../components/Box";
import { parseRawPrice } from "./price";

describe('PriceSummary', () => {

  it('renders without crashing', () => {
    let priceSummaryWrapper = shallow(<PriceSummary/>);
    expect(priceSummaryWrapper).not.toEqual(null);
  })

  describe('when summary is not passed', () => {
    let priceSummaryWrapper;
    beforeAll(()=> {
      priceSummaryWrapper = shallow(<PriceSummary/>);
    })

    it('renders without crashing', () => {
      expect(priceSummaryWrapper).not.toEqual(null);
    })

    it('should render a List container even when summary is not passed as props', () => {
      const listContainer = priceSummaryWrapper.find(Box)
      expect(listContainer).toHaveLength(1)
    })

    it('should not render List when summary is not passed as props', () => {
      const list = priceSummaryWrapper.find(List)
      expect(list).toHaveLength(0)
    })
  });

  describe('when summary is passed', () => {
    let priceSummaryWrapper;
    beforeAll(()=> {
      // const summary = new Map();
      // priceSummaryWrapper = shallow(<PriceSummary summary={summary}/>);
    })

    it('should not render List if summary is empty', () => {
      const summary = new Map();
      priceSummaryWrapper = shallow(<PriceSummary summary={summary}/>);
      const list = priceSummaryWrapper.find(List)
      expect(list).toHaveLength(0)
    })

    it('should render List if summary has 1 item', () => {
      const summary = new Map();
      summary.set("recipeId1",{})
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice="$10.30" shippingPrice={1298}/>);
      const list = priceSummaryWrapper.find(List)
      expect(list).toHaveLength(1)
    })

    it('should render 3 ListItem if summary has 1 item', () => {
      const summary = new Map();
      summary.set("recipeId1",{name : "recipe 1", selected : 2, totalRecipePrice: 1798})
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} />);
      const listItems = priceSummaryWrapper.find(ListItem)
      expect(listItems).toHaveLength(3)
    })

    it('should render 4 ListItem if summary has 2 item', () => {
      const summary = new Map();
      summary.set("recipeId1",{})
      summary.set("recipeId2",{})
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} />);
      const listItems = priceSummaryWrapper.find(ListItem)
      expect(listItems).toHaveLength(4)
    })

    it('should render the correct recipe name ', () => {
      const summary = new Map();
      summary.set("recipeId1",{id: "recipeId1" , name : "recipe 1", selected : 2, totalRecipePrice: 1798 })
      const totalPrice = "$27.98";
      const shippingPrice = 1000;
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice={totalPrice} shippingPrice={shippingPrice}/>);
      const recipeName = summary.get("recipeId1").name;
      expect(priceSummaryWrapper.text()).toContain(recipeName)
    })

    it('should render the correct total recipe price ', () => {
      const summary = new Map();
      summary.set("recipeId1",{id: "recipeId1" , name : "recipe 1", selected : 2, totalRecipePrice: 1798 })
      const totalPrice = "$27.98";
      const shippingPrice = 1000;
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice={totalPrice} shippingPrice={shippingPrice}/>);
      const totalRecipePrice = summary.get("recipeId1").totalRecipePrice;
      expect(priceSummaryWrapper.text()).toContain(parseRawPrice(totalRecipePrice));
    })

    it('should render the correct shipping price', () => {
      const summary = new Map();
      summary.set("recipeId1",{id: "recipeId1" , name : "recipe 1", selected : 2, totalRecipePrice: 1798 })
      const totalPrice = "$27.98";
      const shippingPrice = 1000;
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice={totalPrice} shippingPrice={shippingPrice}/>);
      const shippingPriceListItem = (priceSummaryWrapper.find(ListItem).at(1))
      const shippingPriceListItemAsText = shippingPriceListItem.text()
      expect(shippingPriceListItemAsText).toContain("Shipping " + parseRawPrice(shippingPrice));
    })

    it('should render the correct total box price ', () => {
      const summary = new Map();
      summary.set("recipeId1",{id: "recipeId1" , name : "recipe 1", selected : 2, totalRecipePrice: 1798 })
      const totalPrice = "$27.98";
      const shippingPrice = 1000;
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice={totalPrice} shippingPrice={shippingPrice}/>);
      const totalPriceListItem = (priceSummaryWrapper.find(ListItem).at(2))
      const totalPriceListItemAsText = totalPriceListItem.text()
      expect(totalPriceListItemAsText).toContain("Total " + totalPrice)
    })

    it('should render correct layout for the given summary, totalPrice and shippingCost', () => {
      const summary = new Map();
      summary.set("recipeId1",{id: "recipeId1" , name : "recipe 1", selected : 2, totalRecipePrice: 1798 })
      const totalPrice = "$27.98";
      const shippingPrice = 1000;
      priceSummaryWrapper = shallow(<PriceSummary summary={summary} totalPrice={totalPrice} shippingPrice={shippingPrice}/>);
      expect(priceSummaryWrapper.getElements()).toMatchSnapshot()
    })
  });

});
