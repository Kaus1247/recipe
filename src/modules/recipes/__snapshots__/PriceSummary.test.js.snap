// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceSummary when summary is passed should render correct layout for the given summary, totalPrice and shippingCost 1`] = `
Array [
  <ForwardRef(styled.div)
    width={
      Array [
        "290px",
        "450px",
      ]
    }
  >
    <ForwardRef(styled.ul)
      listStyleType="none"
      margin={0}
      padding="16px"
      width="100%"
    >
      <ForwardRef(styled.li)
        border="none"
        margin={0}
        padding="0px 0px 8px 0px"
        width="100%"
      >
        <ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={9}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-start"
            >
              <ForwardRef(styled.div)
                fontWeight="400"
              >
                 
                recipe 1
                 x 
                2
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={3}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-end"
            >
              <ForwardRef(styled.div)
                fontWeight="400"
              >
                $17.98
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
        </ForwardRef(styled.div)>
      </ForwardRef(styled.li)>
      <ForwardRef(styled.li)
        border="none"
        margin={0}
        padding="0px 0px 8px 0px"
        width="100%"
      >
        <ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={9}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-start"
            >
              <ForwardRef(styled.div)
                fontWeight="400"
              >
                 Shipping 
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={3}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-end"
            >
              <ForwardRef(styled.div)
                fontWeight="400"
              >
                $10.00
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
        </ForwardRef(styled.div)>
      </ForwardRef(styled.li)>
      <ForwardRef(styled.li)
        border="none"
        borderTop="1px solid #E4E4E4"
        margin={0}
        padding="8px 0px 0px 0px"
        width="100%"
      >
        <ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={9}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-start"
            >
              <ForwardRef(styled.div)
                fontWeight="600"
              >
                 Total 
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
          <ForwardRef(styled.div)
            sm={3}
          >
            <ForwardRef(Styled(Box))
              alignItems="center"
              display="flex"
              justifyContent="flex-end"
            >
              <ForwardRef(styled.div)
                fontWeight="600"
              >
                $27.98
              </ForwardRef(styled.div)>
            </ForwardRef(Styled(Box))>
          </ForwardRef(styled.div)>
        </ForwardRef(styled.div)>
      </ForwardRef(styled.li)>
    </ForwardRef(styled.ul)>
  </ForwardRef(styled.div)>,
]
`;
